### DISCUSSION TITLE
Can't output JPEG XL file

### DISCUSSION THREAD

**Discussion by ghost** [ImageMagick/ImageMagick#4825] [Help] [answered]
Hello,

I am trying to convert to JPEG XL file by following commandline with "ImageMagick-7.1.0-23-Q16-HDRI-x64-dll.exe" on Windows 10 21H2 and 21H1 but I can't.

```
"C:\Program Files\ImageMagick\magick.exe" convert test.png test.jxl
```

ImageMagick outputs "test.jxl" with no errors but it is 0 byte. ImageMagick can output other kinds of file. I tested jpg, png, jp2, webp, and avif. 
ImageMagick ver.7.1.0-19 worked well by same command line.
Should I change commandline or install additional software?

Thanks for your time.

**Comment by fmw42** [comment, marked as answer]
Do you have the jpeg-xl delegate library installed. See https://gitlab.com/wg1/jpeg-xl

**Reply by ghost** [reply]
Thanks for your answer and introducing reference software.

I didn't install any jpeg-xl library. But it seems that the library is already built in. ImageMagick can read jpeg-xl files and I can convert from jpeg-xl to any other formats. Following is the result of `magick.exe -version`
```
Version: ImageMagick 7.1.0-23 Q16-HDRI x64 2022-02-05 https://imagemagick.org
Copyright: (C) 1999-2021 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2022 (193030709)
```
ImageMagick 7.1.0-19 can both read and write jpeg-xl files without installing library but I lost it. I'm not sure how to install the library but I have checked that I can convert files with "cjxl.exe" instead.

**Comment by dlemstra** [comment]
I am unable to reproduce your issue on Windows:

```
I:\issues\im7\4825>magick convert logo: test.jxl

I:\issues\im7\4825>dir

10/02/2022  17:47            29.036 test.jxl
               1 File(s)         29.036 bytes

I:\issues\im7\4825>identify test.jxl
test.jxl JXL 640x480 640x480+0+0 8-bit sRGB 29036B 0.031u 0:00.032
```

Can you reproduce this with every image or just with `test.png`?

**Reply by ghost** [reply]
Thanks for your verification.

Yes, every image including "logo:" causes this problem. I have 3 PCs and this problem is reproduced in all of them.
After your posting, I tested it in two Windows10 21H2 PCs, (A) almost no additional software, and (B) some software is installed.
ImageMagick outputs correct jxl files in (A), but 0 byte files in (B).
So not ImageMagick but my PCs except (A) have some problems.
I thought that commonly installed software in 4 PCs caused this problem and I uninstalled them from (B).
But this problem is still reproduced.
I tried "ImageMagick-7.1.0-23-Q8-x64-static.exe" and "ImageMagick-7.1.0-portable-Q16-HDRI-x64.zip" in (B) but didn't improved.

I don't know how to verify this any further. But I found that this is not ImageMagick's problem but mine.

**Reply by dlemstra** [reply]
Maybe some corporate tool is blocking loading of certain libraries?

**Reply by ghost** [reply]
Thank you.
I use ESET internet security Ver 15.0.23.0 in 3 PCs, except (A), including (B).
I disabled ESET and checked, but didn't improve.
I tried safe mode, too. ImageMagick outputted correct PNG file and 0 byte jxl file in safe mode.

**Reply by ghost** [reply]
I tried `magick.exe convert -define jxl:effort=x logo: test.jxl` in 2 PCs including (B).
When `x=1~4`, operation finished instantly and correct jxl files are outputted.
When `x=5~9`, operation took more than 1 second and 0 byte jxl files are outputted.
With using `x=1~4`, 12M pixel large jpg file can be successfully converted to jxl file.

**Reply by fmw42** [reply]
Perhaps your jpeg-ixl delegate version is too old?

**Reply by ghost** [reply]
It may be possible. I couldn't find out what software caused this conflict or brought bad delegate libraries.
I tried Ver 7.1.0-24 but anything wasn't changed.
Anyway, I found out followings in this discussion.

1. Neither this Imagemagick's bug nor my usage, but my PCs' environment prevents Imagemagick from outputing jxl files.
2. By using `-define jxl:effort=1~4` option, this problem can be avioided.
3. "cjxl.exe" isn't affected this problem.

Thanks very much!

**Comment by dlemstra** [comment]
Good news I finally figured out what was causing this and why I could not reproduce the issue. After upgrading the build to VisualStudio 2022 this broke. The compiler does some release optimizations that breaks the jpeg-xl coder. I could not reproduce this because I was testing this with a debug build where the optimization is disabled. This issue should be resolved in the next release of ImageMagick.

**Reply by ghost** [reply]
I have confirmed that Ver. 7.1.0-25 can convert to jxl files successfully.
Thanks a lot.

**Comment by ghost** [comment]
@dlemstra 

Hey,

i am working on Windows 10 Pro 64 22H2. All software is up to date. My system is Xeon E5 2695V3 (14/28C) with 128GB RAM.

```
convert gigapixel.psb output.jxl
convert gigapixel.tiff output.jxl
```
runs some minutes and produces a JXL file with 0 Byte. No crash. The gigapixel file has a size of 60.000 x 40.000 Pixel.

```
Version: ImageMagick 7.1.0-62 Q16-HDRI x64 32ce406:20230212 https://imagemagick.org
Copyright: (C) 1999 ImageMagick Studio LLC
License: https://imagemagick.org/script/license.php
Features: Cipher DPC HDRI Modules OpenCL OpenMP(2.0)
Delegates (built-in): bzlib cairo flif freetype gslib heic jng jp2 jpeg jxl lcms lqr lzma openexr pangocairo png ps raqm raw rsvg tiff webp xml zip zlib
Compiler: Visual Studio 2022 (193431937)
```

How can i help to debug this?

**Comment by snibgo** [comment]
You have IM v7, but use `convert` which uses the old (very old) v6 syntax. I suggest using just `magick` instead (and not `magick convert`).

Does the problem occur with small input files?

When the command is running, how much memory and disk you have free?

**Comment by snibgo** [comment]
Other experiments you can try:
```
magick gigapixel.psb info:
```
This will try to read all pixels into memory, and do some statistics on the values, but no other processing. Does that work?
```
magick -limit memory 1GB -limit map 1GB gigapixel.psb out.jxl
```
[EDIT: move limits to before, instead of after, reading the input.]

This will try to read all pixels into disk instead of memory, which will be slow, but will free memory for the JXL library to use. If that works, then the problem might be the JXL library running out of memory but not correctly noticing the failure.

**Comment by ghost** [comment]
@snibgo 

Thank you very much for your help :-)

`magick gigapixel.psb gigapixel.jxl` did the same, a file 0 byte.

There is no software running so Windows has all resources free (and used by magick)

`magick gigapixel.psb info:` gives
``` batch
gigapixel.psb PSB 59080x39050 59080x39050+0+0 16-bit sRGB 12.8918GiB 0.000u 0:00.023
```
`magick -limit memory 1GB -limit map 1GB gigapixel.psb out.jxl` gives again a file containing 0 bytes.

Is there any way to get a log file about whats going on?

It looks like there is no crash, everything goes down slowly

**Comment by snibgo** [comment]
`magick gigapixel.psb info:` works, so that is good. Try again with "-verbose:"
```
magick gigapixel.psb -verbose info:
```
I expect that will also be okay.

Does the problem occur with small input files?

> Is there any way to get a log file about whats going on?

Insert "`-debug all`" as the first option, eg:
```
magick -debug all -limit memory 1GB -limit map 1GB gigapixel.psb out.jxl
```
I suggest trying a different output format, eg tiff:
```
magick -debug all -limit memory 1GB -limit map 1GB gigapixel.psb out.tiff
```
If that works, then the JXL library (or IM's use of the library) is likely the problem.

I'm just suggesting tests to narrow down the problem. I probably can't fix it, but this may help a developer investigate further.

**Comment by ghost** [comment]
`magick gigapixel.psb -verbose info:` works and gives many stuff

Ok i tried a 10% version of the image this works.


`magick -debug all -limit memory 1GB -limit map 1GB gigapixel.psb out.jxl` runs without specials and ends:

```batch
...
2023-02-23T20:48:35+00:00 2:03.069 124.250u 7.1.0 Cache magick[2604]: cache.c/ReadPixelCachePixels/4720/Cache
  gigapixel.psb[0][59080x1+0+35490]
2023-02-23T20:48:36+00:00 2:03.464 124.625u 7.1.0 Cache magick[2604]: cache.c/ReadPixelCachePixels/4720/Cache
  gigapixel.psb[0][59080x1+0+35880]
2023-02-23T20:48:36+00:00 2:03.857 125.016u 7.1.0 Cache magick[2604]: cache.c/ReadPixelCachePixels/4720/Cache
  gigapixel.psb[0][59080x1+0+36270]
2023-02-23T20:48:37+00:00 2:04.247 125.406u 7.1.0 Cache magick[2604]: cache.c/ReadPixelCachePixels/4720/Cache
  gigapixel.psb[0][59080x1+0+36660]
2023-02-23T20:48:37+00:00 2:04.638 125.797u 7.1.0 Cache magick[2604]: cache.c/ReadPixelCachePixels/4720/Cache
  gigapixel.psb[0][59080x1+0+37050]
2023-02-23T20:48:37+00:00 2:05.029 126.188u 7.1.0 Cache magick[2604]: cache.c/ReadPixelCachePixels/4720/Cache
  gigapixel.psb[0][59080x1+0+37440]
2023-02-23T20:48:38+00:00 2:05.421 126.563u 7.1.0 Cache magick[2604]: cache.c/ReadPixelCachePixels/4720/Cache
  gigapixel.psb[0][59080x1+0+37830]
2023-02-23T20:48:38+00:00 2:05.810 126.938u 7.1.0 Cache magick[2604]: cache.c/ReadPixelCachePixels/4720/Cache
  gigapixel.psb[0][59080x1+0+38220]
2023-02-23T20:48:38+00:00 2:06.199 127.313u 7.1.0 Cache magick[2604]: cache.c/ReadPixelCachePixels/4720/Cache
  gigapixel.psb[0][59080x1+0+38610]
2023-02-23T20:48:39+00:00 2:06.613 127.719u 7.1.0 Cache magick[2604]: cache.c/ReadPixelCachePixels/4720/Cache
  gigapixel.psb[0][59080x1+0+39000]
2023-02-23T20:48:39+00:00 2:06.666 127.766u 7.1.0 Cache magick[2604]: cache.c/ReadPixelCachePixels/4720/Cache
  gigapixel.psb[0][59080x1+0+39049]

C:\Users\..\>
```

what gives again a file containing 0 bytes.

`magick -debug all -limit memory 1GB -limit map 1GB gigapixel.psb out.tiff` crashes with

```batch
...
magick: Maximum TIFF file size exceeded. `TIFFAppendToStrip' @ error/tiff.c/TIFFErrors/598.
```
(...thats correct so far...)

**Comment by ghost** [comment]
> Ok i tried a 10% version of the image this works.

Oh no. Stop - this was another image from yesterday. So i must test this first (maybe, this one works).
